presto-native-execution/presto_cpp/main/expression/RowExpressionEvaluator.cpp

```diff
@@ -0,0 +1,419 @@
+      res["type"] = expr->type()->toString();
+    } else {
+      // divide(0, 0) is a constant expression with constant inputs.
+      res = inputRowExpr;
+    }
+  } else if (
+      auto field =
+          std::dynamic_pointer_cast<const exec::FieldReference>(expr)) {
+    // variable
+    res = toVariableReferenceExpression(field.get());
+  } else if (expr->isSpecialForm()) {
+    // special
+    res = getSpecialForm(expr, inputRowExpr);
+  } else if (auto func = expr->vectorFunction()) {
+    // Check if special form expression or call expression.
+    std::unordered_set<std::string> specialForms = {
+        "in", "cast", "try", "is_null", "row_constructor"};
+    if (specialForms.count(expr->name()) != 0) {
+      // special
+      res = getSpecialForm(expr, inputRowExpr);
+    } else {
+      // call
+      res["@type"] = "call";
+      protocol::Signature signature;
+      std::string exprName = expr->name();
+      if (scalarOperatorMap_.find(expr->name()) != scalarOperatorMap_.end()) {
+        exprName = scalarOperatorMap_.at(expr->name());
+      }
+      signature.name = exprName;
+      res["displayName"] = exprName;
+      signature.kind = protocol::FunctionKind::SCALAR;
+      signature.typeVariableConstraints = {};
+      signature.longVariableConstraints = {};
+      signature.returnType = expr->type()->toString();
+
+      std::vector<protocol::TypeSignature> argumentTypes;
+      auto inputs = expr->inputs();
+      auto numArgs = inputs.size();
+      argumentTypes.reserve(numArgs);
+      for (auto i = 0; i < numArgs; i++) {
+        argumentTypes.emplace_back(inputs[i]->type()->toString());
+      }
+      signature.argumentTypes = argumentTypes;
+      signature.variableArity = false;
+
+      protocol::BuiltInFunctionHandle builtInFunctionHandle;
+      builtInFunctionHandle._type = "$static";
+      builtInFunctionHandle.signature = signature;
+      res["functionHandle"] = builtInFunctionHandle;
+      res["returnType"] = expr->type()->toString();
+      for (const auto& input : inputs) {
+        res["arguments"].push_back(veloxExprToRowExpr(input, inputRowExpr));
+      }
+    }
+  } else {
+    VELOX_USER_FAIL(
+        "Unable to convert Velox Expr to Presto RowExpression: {}",
+        expr->toString());
+  }
+
+  return res;
+}
+
+// Compiles protocol::RowExpression into a velox expression, with constant
+// folding enabled.
+std::shared_ptr<const velox::exec::Expr>
+RowExpressionEvaluator::compileExpression(
+    const std::shared_ptr<protocol::RowExpression>& inputRowExpr) {
+  auto typedExpr = exprConverter_.toVeloxExpr(inputRowExpr);
+  exec::ExprSet exprSet{{typedExpr}, execCtx_.get()};
+  auto compiledExprs =
+      exec::compileExpressions({typedExpr}, execCtx_.get(), &exprSet, true);
+  return compiledExprs[0];
+}
+
+void RowExpressionEvaluator::evaluateExpression(
+    const std::vector<std::unique_ptr<folly::IOBuf>>& body,
+    proxygen::ResponseHandler* downstream) {
+  try {
```

Comment:
Might be good to abstract a function to return a json::array from folly::IOBuf and make this function take a json array and return a json for result.

On second thoughts maybe such a function to return a json::array from folly::IOBuf exists or can be used elsewhere. It would be a common utility.

Reply:
Thanks for the suggestion, added helper functions and updated accordingly. I could not find a similar utility function in the codebase. Could you please take another look?